src/Fixer/ClassNotation/FinalStaticAccessFixer.php

```diff
@@ -0,0 +1,153 @@
+use PhpCsFixer\Tokenizer\Tokens;
+
+/**
+ * @author ntzm
+ */
+final class FinalStaticAccessFixer extends AbstractFixer
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function getDefinition()
+    {
+        return new FixerDefinition(
+            'Converts `static` access to `self` access in final classes.',
+            [
+                new CodeSample(
+                    '<?php
+final class Sample
+{
+    public function getFoo()
+    {
+        return static::class;
+    }
+}
+'
+                ),
+            ]
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getPriority()
+    {
+        // Should be run after FinalInternalClass and PhpUnitTestCaseStaticMethodCalls
+        return -1;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isCandidate(Tokens $tokens)
+    {
+        return $tokens->isAllTokenKindsFound([T_FINAL, T_CLASS, T_STATIC]);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function applyFix(\SplFileInfo $file, Tokens $tokens)
+    {
+        for ($index = $tokens->count() - 1; 0 <= $index; --$index) {
+            if (!$tokens[$index]->isGivenKind(T_FINAL)) {
+                continue;
+            }
+
+            $classTokenIndex = $tokens->getNextMeaningfulToken($index);
+
+            if (!$tokens[$classTokenIndex]->isGivenKind(T_CLASS)) {
+                continue;
+            }
+
+            $startClassIndex = $tokens->getNextTokenOfKind(
+                $classTokenIndex,
+                ['{']
+            );
+
+            $endClassIndex = $tokens->findBlockEnd(
+                Tokens::BLOCK_TYPE_CURLY_BRACE,
+                $startClassIndex
+            );
+
+            $this->replaceStaticAccessWithSelfAccessBetween(
+                $tokens,
+                $startClassIndex,
+                $endClassIndex
+            );
+        }
+    }
+
+    /**
+     * @param int $startIndex
+     * @param int $endIndex
+     */
+    private function replaceStaticAccessWithSelfAccessBetween(
+        Tokens $tokens,
+        $startIndex,
+        $endIndex
+    ) {
+        for ($index = $startIndex; $index <= $endIndex; ++$index) {
+            if ($tokens[$index]->isGivenKind(T_CLASS)) {
+                $index = $this->getEndOfAnonymousClass($tokens, $index);
+
+                continue;
+            }
+
+            if (!$tokens[$index]->isGivenKind(T_STATIC)) {
+                continue;
+            }
+
+            $doubleColonIndex = $tokens->getNextMeaningfulToken($index);
+
+            if (!$tokens[$doubleColonIndex]->isGivenKind(T_DOUBLE_COLON)) {
+                continue;
+            }
+
+            $tokens[$index] = new Token([T_STRING, 'self']);
+        }
+    }
+
+    /**
+     * @param int $index
+     *
+     * @return int
+     */
+    private function getEndOfAnonymousClass(Tokens $tokens, $index)
+    {
+        $instantiationBraceStart = $tokens->getNextMeaningfulToken($index);
```

Comment:
can't you jump directly to `$bodyBraceStart = $tokens->getNextTokenOfKind($index, ['{']);` ? I think an anonymous always have a body  (`{` `}`) and you are never interested in the tokens between `new class` and the end `}`

Reply:
I tried that! See the other test cases I added, but imagine this:

```php
new class (function () {}) {};
```

Reply:
Very nice catch!
i wonder if we've more places where this case was not considered

Reply:
I was wondering that, it's definitely worth having a look